UnknownCurrencyException(CurrencyCode=EUR)

How can I resolve this issue? This happens with versions 1.0 and 1,1 on Java 8u111 and higher.

When I use 

```
CurrencyUnit currency = Monetary.getCurrency(valuta);
```

I get an 

> UnknownCurrencyException(CurrencyCode=EUR)

Can this be a timing issue? Is the JDKCurrencyProvider lazy loaded?

With `moneta-bp:1.1` stacktrace is slightly different (note `BaseMonetaryCurrenciesSingletonSpi` vs `DefaultMonetaryCurrenciesSingletonSpi`):
```
java.util.concurrent.ExecutionException: UnknownCurrencyException [currencyCode=EUR]
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at com.github.ayastrebov.unknown.currency.Main.main(Main.java:21)
Caused by: UnknownCurrencyException [currencyCode=EUR]
	at org.javamoney.moneta.spi.base.BaseMonetaryCurrenciesSingletonSpi.getCurrency(BaseMonetaryCurrenciesSingletonSpi.java:53)
	at javax.money.Monetary.getCurrency(Monetary.java:150)
	at com.github.ayastrebov.unknown.currency.Main$Task.call(Main.java:35)
	at com.github.ayastrebov.unknown.currency.Main$Task.call(Main.java:31)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

and also have got different kind of error once:
```
java.util.concurrent.ExecutionException: java.util.ServiceConfigurationError: javax.money.spi.CurrencyProviderSpi: Provider org.javamoney.moneta.internal.JDKCurrencyProvider could not be instantiated
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at com.github.ayastrebov.unknown.currency.Main.main(Main.java:21)
Caused by: java.util.ServiceConfigurationError: javax.money.spi.CurrencyProviderSpi: Provider org.javamoney.moneta.internal.JDKCurrencyProvider could not be instantiated
	at java.util.ServiceLoader.fail(ServiceLoader.java:232)
	at java.util.ServiceLoader.access$100(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:384)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at org.javamoney.moneta.internal.PriorityAwareServiceProvider.loadServices(PriorityAwareServiceProvider.java:117)
	at org.javamoney.moneta.internal.PriorityAwareServiceProvider.getServices(PriorityAwareServiceProvider.java:75)
	at javax.money.spi.Bootstrap.getServices(Bootstrap.java:109)
	at org.javamoney.moneta.internal.DefaultMonetaryCurrenciesSingletonSpi.getDefaultProviderChain(DefaultMonetaryCurrenciesSingletonSpi.java:112)
	at org.javamoney.moneta.internal.DefaultMonetaryCurrenciesSingletonSpi.collectProviders(DefaultMonetaryCurrenciesSingletonSpi.java:69)
	at org.javamoney.moneta.internal.DefaultMonetaryCurrenciesSingletonSpi.getCurrencies(DefaultMonetaryCurrenciesSingletonSpi.java:42)
	at org.javamoney.moneta.spi.base.BaseMonetaryCurrenciesSingletonSpi.getCurrency(BaseMonetaryCurrenciesSingletonSpi.java:50)
	at javax.money.Monetary.getCurrency(Monetary.java:150)
	at com.github.ayastrebov.unknown.currency.Main$Task.call(Main.java:35)
	at com.github.ayastrebov.unknown.currency.Main$Task.call(Main.java:31)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: java.util.HashMap$Node cannot be cast to java.util.HashMap$TreeNode
	at java.util.HashMap$TreeNode.moveRootToFront(HashMap.java:1827)
	at java.util.HashMap$TreeNode.treeify(HashMap.java:1944)
	at java.util.HashMap.treeifyBin(HashMap.java:771)
	at java.util.HashMap.putVal(HashMap.java:643)
	at java.util.HashMap.put(HashMap.java:611)
	at org.javamoney.moneta.internal.JDKCurrencyProvider.<init>(JDKCurrencyProvider.java:43)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)
	... 16 more
```

@keilw 

> That is why I was questioning the dependency

I can reproduce it with 
```xml
    <dependencies>
        <dependency>
            <groupId>org.javamoney</groupId>
            <artifactId>moneta</artifactId>
            <version>1.1</version>
        </dependency>
    </dependencies>
```
```
[INFO] --- maven-dependency-plugin:2.1:tree (default-cli) @ unknown-currency ---
[INFO] com.github.ayastrebov:unknown-currency:jar:1.0.0-SNAPSHOT
[INFO] \- org.javamoney:moneta:jar:1.1:compile
[INFO]    +- javax.money:money-api:jar:1.0.1:compile
[INFO]    \- javax.annotation:javax.annotation-api:jar:1.2:compile
```
```
java.util.concurrent.ExecutionException: UnknownCurrencyException [currencyCode=EUR]
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at com.github.ayastrebov.unknown.currency.Main.main(Main.java:21)
Caused by: UnknownCurrencyException [currencyCode=EUR]
	at javax.money.spi.MonetaryCurrenciesSingletonSpi.getCurrency(MonetaryCurrenciesSingletonSpi.java:74)
	at javax.money.Monetary.getCurrency(Monetary.java:422)
	at com.github.ayastrebov.unknown.currency.Main$Task.call(Main.java:35)
	at com.github.ayastrebov.unknown.currency.Main$Task.call(Main.java:31)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)

```

>  let's discuss the Java 7 issues in the ri-bp ticket now.

Ok, though I am sure the core issue is the same

--- a/moneta-core/src/main/java/org/javamoney/moneta/internal/JDKCurrencyProvider.java
+++ b/moneta-core/src/main/java/org/javamoney/moneta/internal/JDKCurrencyProvider.java
@@ -26,7 +26,7 @@ import javax.money.spi.CurrencyProviderSpi;
 /**
  * Default implementation of a {@link CurrencyUnit} based on the using the JDK's
  * {@link Currency}.
- * 
+ *
  * @version 0.5.1
  * @author Anatole Tresch
  * @author Werner Keil
@@ -34,14 +34,17 @@ import javax.money.spi.CurrencyProviderSpi;
 public class JDKCurrencyProvider implements CurrencyProviderSpi {
 
 	/** Internal shared cache of {@link javax.money.CurrencyUnit} instances. */
-	private static final Map<String, CurrencyUnit> CACHED = new HashMap<>();
+    private static final Map<String, CurrencyUnit> CACHED = loadCurrencies();
 
-	public JDKCurrencyProvider() {
-		for (Currency jdkCurrency : Currency.getAvailableCurrencies()) {
-			CurrencyUnit cu = new JDKCurrencyAdapter(jdkCurrency);
-			CACHED.put(cu.getCurrencyCode(), cu);
-		}
-	}
+    private static Map<String, CurrencyUnit> loadCurrencies() {
+        Set<Currency> availableCurrencies = Currency.getAvailableCurrencies();
+        Map<String, CurrencyUnit> result = new HashMap<>(availableCurrencies.size());
+        for (Currency jdkCurrency : availableCurrencies) {
+            CurrencyUnit cu = new JDKCurrencyAdapter(jdkCurrency);
+            result.put(cu.getCurrencyCode(), cu);
+        }
+        return Collections.unmodifiableMap(result);
+    }
 
     @Override
     public String getProviderName(){

